File: package/umt_python/src/math/percentile.py
import math
from collections.abc import Sequence


def percentile(array: Sequence[float], pct: float) -> float:
    """
    Calculate the nth percentile of values in an array.

    Args:
        array: Array of numbers.
        pct: Percentile value (0-100).

    Returns:
        The percentile value.

    Raises:
        ValueError: If percentile is not between 0 and 100.

    Example:
        >>> percentile([1, 2, 3, 4, 5], 50)
        3
        >>> percentile([1, 2, 3, 4, 5], 25)
        2
        >>> percentile([1, 2, 3, 4, 5], 75)
        4
    """
    if len(array) == 0:
        return float("nan")

    if pct < 0 or pct > 100:
        raise ValueError("Percentile must be between 0 and 100")

    sorted_array = sorted(array)
    index = (pct / 100) * (len(sorted_array) - 1)
    lower_index = math.floor(index)
    upper_index = math.ceil(index)

    if lower_index == upper_index:
        return sorted_array[lower_index]

    lower_value = sorted_array[lower_index]
    upper_value = sorted_array[upper_index]
    weight = index - lower_index

    return lower_value + (upper_value - lower_value) * weight

File: package/main/src/Math/percentile.ts
import { quickSort } from "@/Array/quickSort";

/**
 * Calculate the nth percentile of values in an array
 * @param array - Array of numbers
 * @param percentile - Percentile value (0-100)
 * @returns The percentile value
 * @example
 * percentile([1, 2, 3, 4, 5], 50); // 3 (50th percentile - median)
 * percentile([1, 2, 3, 4, 5], 25); // 2 (25th percentile)
 * percentile([1, 2, 3, 4, 5], 75); // 4 (75th percentile)
 */
export const percentile = (array: number[], percentile: number): number => {
  if (array.length === 0) {
    return Number.NaN;
  }

  const sortedArray = quickSort([...array]);
  const index = (percentile / 100) * (sortedArray.length - 1);
  const lowerIndex = Math.floor(index);
  const upperIndex = Math.ceil(index);

  if (lowerIndex === upperIndex) {
    return sortedArray[lowerIndex];
  }

  const lowerValue = sortedArray[lowerIndex];
  const upperValue = sortedArray[upperIndex];
  const weight = index - lowerIndex;

  return lowerValue + (upperValue - lowerValue) * weight;
};
